File: zenoh/examples/zenoh-net/zn_pub_thr.rs
//
// Copyright (c) 2017, 2020 ADLINK Technology Inc.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ADLINK zenoh team, <zenoh@adlink-labs.tech>
//
use clap::{App, Arg};
use zenoh::net::ResKey::*;
use zenoh::net::*;
use zenoh::Properties;

fn main() {
    // initiate logging
    env_logger::init();
    let (config, size) = parse_args();

    let data: ZBuf = (0usize..size)
        .map(|i| (i % 10) as u8)
        .collect::<Vec<u8>>()
        .into();

    let session = open(config.into()).wait().unwrap();

    let reskey = RId(session
        .declare_resource(&RName("/test/thr".to_string()))
        .wait()
        .unwrap());
    let _publ = session.declare_publisher(&reskey).wait().unwrap();

    loop {
        session.write_ext(
            &reskey,
            data.clone(),
            encoding::DEFAULT,
            data_kind::DEFAULT,
            CongestionControl::Block, // Make sure to not drop messages because of congestion control
        );
    }
}

fn parse_args() -> (Properties, usize) {
    let args = App::new("zenoh-net throughput pub example")
        .arg(
            Arg::from_usage("-m, --mode=[MODE] 'The zenoh session mode (peer by default).")
                .possible_values(&["peer", "client"]),
        )
        .arg(Arg::from_usage(
            "-e, --peer=[LOCATOR]...  'Peer locators used to initiate the zenoh session.'",
        ))
        .arg(Arg::from_usage(
            "-l, --listener=[LOCATOR]...   'Locators to listen on.'",
        ))
        .arg(Arg::from_usage(
            "-c, --config=[FILE]      'A configuration file.'",
        ))
        .arg(Arg::from_usage(
            "<PAYLOAD_SIZE>          'Sets the size of the payload to publish'",
        ))
        .get_matches();

    let mut config = if let Some(conf_file) = args.value_of("config") {
        Properties::from(std::fs::read_to_string(conf_file).unwrap())
    } else {
        Properties::default()
    };
    for key in ["mode", "peer", "listener"].iter() {
        if let Some(value) = args.values_of(key) {
            config.insert(key.to_string(), value.collect::<Vec<&str>>().join(","));
        }
    }
    if args.is_present("no-multicast-scouting") {
        config.insert("multicast_scouting".to_string(), "false".to_string());
    }

    let size = args
        .value_of("PAYLOAD_SIZE")
        .unwrap()
        .parse::<usize>()
        .unwrap();

    (config, size)
}

File: zenoh-ffi/examples/zn_pub_thr.c
/*
 * Copyright (c) 2017, 2020 ADLINK Technology Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
 * which is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Contributors:
 *   ADLINK zenoh team, <zenoh@adlink-labs.tech>
 */
#include "zenoh-ffi.h"
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>

int main(int argc, char** argv) {
  char *locator = 0;
  if (argc < 2) {
    printf("USAGE:\n\tzn_pub_thr <payload-size> [<zenoh-locator>]\n\n");
    exit(-1);
  }
  size_t len = atoi(argv[1]);  
  printf("Running throughput test for payload of %zu bytes.\n", len);
  if (argc > 2) {
    locator = argv[2];
  }  

  ZNSession *s = zn_open(PEER, locator, 0);
  if (s == 0) {
    printf("Unable to open session!\n");
    exit(-1);
  }

  char *data = (char*) malloc(len);
  memset(data, 1, len);

  size_t rid = zn_declare_resource(s, "/test/thr");
  while (1) {
    zn_write_wrid(s, rid, data, len);
  }
}